qiskit_aqua/utils/boolean_logic.py

```diff
@@ -0,0 +1,231 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2019 IBM.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+"""
+The Boolean Logic Utility Classes.
+"""
+
+import itertools
+import logging
+from abc import abstractmethod, ABC
+
+
+from qiskit import QuantumCircuit, QuantumRegister
+from qiskit.qasm import pi
+
+from .mct import mct
+
+logger = logging.getLogger(__name__)
+
+
+def _or(clause_expr, clause_index, circuit, qr_variable, qr_clause, qr_ancilla, mct_mode):
+    qs = [abs(v) for v in clause_expr]
+    ctl_bits = [qr_variable[idx - 1] for idx in qs]
+    anc_bits = [qr_ancilla[idx] for idx in range(len(qs) - 2)] if qr_ancilla else None
+    tgt_bits = qr_clause[clause_index]
+    for idx in [v for v in clause_expr if v > 0]:
+        circuit.u3(pi, 0, pi, qr_variable[idx - 1])
+    circuit.mct(ctl_bits, tgt_bits, anc_bits, mode=mct_mode)
+    for idx in [v for v in clause_expr if v > 0]:
+        circuit.u3(pi, 0, pi, qr_variable[idx - 1])
+
+
+def _and(clause_expr, clause_index, circuit, qr_variable, qr_clause, qr_ancilla, mct_mode):
+    qs = [abs(v) for v in clause_expr]
+    ctl_bits = [qr_variable[idx - 1] for idx in qs]
+    anc_bits = [qr_ancilla[idx] for idx in range(len(qs) - 2)] if qr_ancilla else None
+    tgt_bits = qr_clause[clause_index]
+    circuit.mct(ctl_bits, tgt_bits, anc_bits, mode=mct_mode)
+
+
+def _set_up_register(num_qubits_needed, provided_register, description):
+    if provided_register is None:
+        if num_qubits_needed > 0:
+            return QuantumRegister(num_qubits_needed, name=description[0])
+    else:
+        num_qubits_provided = len(provided_register)
+        if num_qubits_needed > num_qubits_provided:
+            raise ValueError(
+                'The {} QuantumRegister needs {} qubits, but the provided register contains only {}.'.format(
+                    description, num_qubits_needed, num_qubits_provided
+                ))
+        else:
+            if num_qubits_needed < num_qubits_provided:
+                logger.warning(
+                    'The {} QuantumRegister only needs {} qubits, but the provided register contains {}.'.format(
+                        description, num_qubits_needed, num_qubits_provided
+                    ))
+            return provided_register
+
+
+class BooleanLogicNormalForm(ABC):
+    def __init__(self, cnf_expr):
```

Comment:
could we have docstring to describe the arguments?

Reply:
sure will do